### PR TITLE
testing: Use t.cleanup in TestAgent , and fix two flaky tests

### PR DESCRIPTION
**Pull request by dnephin**
While working on another change I had to make some changes to TestAgent, so I am splitting out some of these changes ahead of the other PR.

While testing these changes I hit a few flaky tests, so this PR includes fixes for them, in separate commits. Details of each change in the commit messages.

### CODE REVIEW

File: agent/dns_test.go [hunk 1]
```diff
@@ -5829,16 +5829,12 @@ func TestDNS_NonExistentDC_RPC(t *testing.T) {
...
 
 	// Join LAN cluster
 	addr := fmt.Sprintf("127.0.0.1:%d", s.Config.SerfPortLAN)
 	_, err := c.JoinLAN([]string{addr})
 	require.NoError(t, err)
-	retry.Run(t, func(r *retry.R) {
-		require.Len(r, s.LANMembers(), 2)
-		require.Len(r, c.LANMembers(), 2)
-	})
+	testrpc.WaitForTestAgent(t, c.RPC, "dc1")
```
Comment (dnephin) [inline agent/dns_test.go:5837]: This seems to work, but I'm not confident it is the correct function to use. Is there a better option? 

The flake was `TestAgent.dns: rpc error: error=No known Consul servers`, so I guess we have to wait for the client to fully join the cluster and receive the RPC address of the servers (or something like that) ?

File: agent/acl_test.go [hunk 2]
```diff
@@ -52,11 +36,15 @@ type TestACLAgent struct {
 // Basically it needs a local state for some of the vet* functions, a logger and a delegate.
 // The key is that we are the delegate so we can control the ResolveToken responses
 func NewTestACLAgent(t *testing.T, name string, hcl string, resolveAuthz authzResolver, resolveIdent identResolver) *TestACLAgent {
-	a := &TestACLAgent{Name: name, HCL: hcl, resolveAuthzFn: resolveAuthz, resolveIdentFn: resolveIdent}
-	dataDir := `data_dir = "acl-agent"`
+	a := &TestACLAgent{resolveAuthzFn: resolveAuthz, resolveIdentFn: resolveIdent}
+
+	dataDir := testutil.TempDir(t, "acl-agent")
```
Comment (dnephin) [inline agent/acl_test.go:41]: I wonder if we should have `testutil.TempDir` call `t.Cleanup()` itself, removing the need for all the callers to add it. Thoughts?

File: agent/testagent.go [hunk 3]
```diff
@@ -274,7 +216,6 @@ func (a *TestAgent) Start(t *testing.T) (err error) {
 	a.Agent.StartSync()
 
 	if err := a.waitForUp(); err != nil {
-		cleanupTmpDir()
 		a.Shutdown()
```
Comment (dnephin) [inline agent/testagent.go:219]: In a follow up we could register this `Shutdown` using `t.Cleanup` as well, and remove the need to call it in every test.

File: agent/acl_test.go (hunk 2 above)
Comment (i0rek) [reply agent/acl_test.go:41]: I like that! I don't know of anything that would speak against it.

File: agent/dns_test.go (hunk 1 above)
Comment (i0rek) [reply agent/dns_test.go:5837]: `WaitForTestAgent` queries the catalog, which will only return once there is a leader so I think it does exactly what you wanted.

File: agent/testagent.go [hunk 4]
```diff
@@ -155,35 +147,14 @@ func (a *TestAgent) Start(t *testing.T) (err error) {
 		name = "TestAgent"
 	}
 
-	var cleanupTmpDir = func() {
```
Comment (i0rek) [inline agent/testagent.go:158]: I don't see where the data dir is cleaned up now.

Comment (dnephin) [reply agent/testagent.go:158]: Oh ya, good catch! I forgot to add the `t.Cleanup`!  I'll add it to `testutil.TempDir`

**Pull request merged**